File: priority-queue/heap.py
class MaxHeap:
    def __init__(self):
        super().__init__()
        self.a = []

    def sift_down(self, i):
        while 2 * i + 1 < len(self.a):
            left = 2 * i + 1
            right = 2 * i + 2
            j = left

            if right < len(self.a) and self.a[right] > self.a[left]:
                j = right
            if self.a[i] > self.a[j]:
                break
            self.a[i], self.a[j] = self.a[j], self.a[i]
            i = j

    def sift_up(self, i):
        while (i - 1) // 2 >= 0 and self.a[i] > self.a[(i - 1) // 2]:
            self.a[i], self.a[(i - 1) // 2] = self.a[(i - 1) // 2], self.a[i]
            i = (i - 1) // 2

    def extract_max(self):
        max_el = self.a[0]
        self.a[0] = self.a[-1]
        self.a.pop(-1)
        self.sift_down(0)
        return max_el

    def insert(self, key):
        self.a.append(key)
        self.sift_up(len(self.a) - 1)


def main():
    n = int(input())
    heap = MaxHeap()
    for _ in range(n):
        inp = input()
        print(heap.extract_max()) if inp == "GET" else heap.insert(int(inp))


if __name__ == '__main__':
    main()

File: priority-queue/Heap.java
import java.util.ArrayList;
import java.util.List;

public class Heap {
    private int[] heap;
    private int size;
    private int maxSize;

    public Heap(int maxSize){
        this.maxSize = maxSize;
        this.size = 0;
        this.heap = new int[maxSize];
    }

    public void add(int value){
        size++;
        heap[size - 1] = value;
        siftUp(size - 1);
    }

    public int get(){
        int value = heap[0];
        heap[0] = heap[size - 1];
        size--;
        siftDown(0);
        return value;
    }

    public void siftUp(int i){
        while(heap[i] > heap[(i - 1)/2]){
            swap(i, (i - 1)/2);
            i = (i - 1)/2;
        }
    }

    public void siftDown(int i){
        while(2 * i + 1 < size){
            int left = 2 * i + 1;
            int right = 2 * i + 2;
            int j = left;
            if(right < size && heap[right] > heap[left])
                j = right;
            if(heap[i] >= heap[j])
                break;
            swap(i, j);
            i = j;
        }
    }

    private void swap(int i1, int i2){
        int tmp = heap[i1];
        heap[i1] = heap[i2];
        heap[i2] = tmp;
    }

}